File: observatory/codex-catalog.mjs
import { execFile } from 'node:child_process';
import { promisify } from 'node:util';
import { join, resolve } from 'node:path';
import { homedir } from 'node:os';
import { fileURLToPath } from 'node:url';
const exec = promisify(execFile);
export async function readCodexCatalog(project, { database = join(homedir(), '.codex/state_5.sqlite'), python = 'python3', roots = [project], includeArchived = false } = {}) {
  const allowed = [...new Set(roots.map(root => resolve(root)))];
  const { stdout } = await exec(python, [fileURLToPath(new URL('./codex-catalog.py', import.meta.url)), database, JSON.stringify(allowed), includeArchived ? 'all' : 'active'], { timeout: 5000, maxBuffer: 1024 * 1024 });
  const rows = JSON.parse(stdout);
  if (!Array.isArray(rows) || rows.some(row => typeof row.id !== 'string' || !allowed.includes(row.cwd))) throw new Error('Unsupported Codex catalog');
  return rows;
}

File: observatory/codex-catalog.py
"""Read only the current project's user-task catalog; never read message bodies."""
import json
import pathlib
import sqlite3
import sys

database = pathlib.Path(sys.argv[1]).resolve()
projects = json.loads(sys.argv[2])
if not isinstance(projects, list) or not projects or any(not isinstance(p, str) for p in projects):
    raise ValueError('Expected verified project roots')
connection = sqlite3.connect(database.as_uri() + '?mode=ro', uri=True, timeout=1)
connection.row_factory = sqlite3.Row
placeholders = ','.join('?' for _ in projects)
rows = connection.execute('''SELECT id, cwd, COALESCE(name, title) AS title,
    updated_at_ms, model, reasoning_effort, archived
    FROM threads WHERE cwd IN (''' + placeholders + ''') AND (archived = 0 OR ? = 1) AND thread_source = 'user'
    ORDER BY updated_at DESC''', projects + [int(len(sys.argv) > 3 and sys.argv[3] == 'all')])
print(json.dumps([dict(row) for row in rows]))
